### PR TITLE
DPL: add metric for consumed timeframes

### PR DESCRIPTION
**Pull request by ktf**
In order for this to work correctly a few things are needed.

* Outputs which is are not created on a timeframe basis needs to be marked
  as Lifetime::QA
* For all the timeframe sinks, we produce automatically a message DPL/SUMMARY/<unique id for a given sink>
  which marks that (part of) a timeframe has reached that sink.
* All those DPL/SUMMARY outputs are garbage collected by the
  dummy-sink, since they are dangling. Whenever all of them arrives, we
  increase the counter of consumed timeframes.
* Notice that in order to avoid accumulating stuff on the dummy sink,
  we introduced a new CompletionOp::ConsumeExisting which is like process,
  but it will drop the payload (but not the header) of a given O2 message pair.

### CODE REVIEW

**Comment by davidrohr** [comment]
Haven't checked yet, but what would be good in combination would be a generic checker, that could be placed in sources, and that would count how many TFs the source has sent, and how many have arrived, via whatever out-of-band channel to the dummy sink.
This could be simple class which could be instantiated in the ctf-reader / raw-file-reader / raw-data-proxy / tfreader. I think in all of them we need the same throttling mechanism.
We could also produce the warnings in that class, such that all the sources would send corresponding warnings, in case the processing chain runs full. I think these warnings will be quite important for the shifters eventually, so we should tune them somewhat to give no message flood, but give a good alert in case something goes wrong.

**Comment by ktf** [comment]
Yes, that was my idea. Right now I am implementing it as a service for the AOD reader, but indeed it could be a generic one. My idea would be that rather than using "Lifetime::Enumeration" for the source we could use "Lifetime::TimeframeLimited" so that the enumeration happens only within the limit of some "maximum number of timeframes in fly".

**Comment by davidrohr** [comment]
For reference, as already written on Mattermost, this PR breaks the full system test. I see that only the TPC does processing, the other detectors' processes do not receive data.

**Review by davidrohr — DISMISSED** [review]
seems to work now, at least it is not breaking the FST on the EPN anymore

**Comment by ktf** [comment]
Ok, I still see something unexpected (consumed timeframes not reported in one of my analysis tests, while the actual processing works) so I might need a bit more time. I also have an idea on how to have the feedback look without the driver.

**Comment by davidrohr** [comment]
> I also have an idea on how to have the feedback look without the driver.

Great, please check also what I have written on the Jira. Particularly, we should have some info messages when throttling starts / stops.

**Comment by ktf** [comment]
Updated with proper handling of multiple sinks. With this I get reasonable results also for the analysis tests.

**Comment by ktf** [comment]
The broken tests need to be updated, since they test the topology and the topology has changed. Apart from it, I cannot find any regression and the metric is always there and looks correct.

**Comment by ktf** [comment]
Test fixed.

File: Framework/Core/src/DataRelayer.cxx
Comment (ktf) [inline Framework/Core/src/DataRelayer.cxx:614]: @matthiasrichter I suspect this will have to change to make it working with the optimized DataModel

File: Framework/Core/src/DataRelayer.cxx
Comment (ktf) [inline Framework/Core/src/DataRelayer.cxx]: @matthiasrichter same here. I use the presence of the header (with a null payload) to indicate that part of the message has been already consumed.

**Comment by davidrohr** [comment]
I ran this through the full system test on an EPN, and it doesn't seem to break the test, but there are several error messages:
```
[549981:ctf-writer]: [17:16:07][ERROR] SendingMessage
[549982:internal-dpl-injected-dummy-sink]: [17:16:07][ERROR] Invoked
[549982:internal-dpl-injected-dummy-sink]: [17:16:07][ERROR] sink:false action:consumeExisting
[549982:internal-dpl-injected-dummy-sink]: [17:16:07][ERROR] ConsumeExisting
[549938:mch-data-decoder]: [17:16:07][ERROR] sink:false action:consume
[549982:internal-dpl-injected-dummy-sink]: [17:16:07][ERROR] Invoked
[549982:internal-dpl-injected-dummy-sink]: [17:16:07][ERROR] sink:false action:consumeExisting
[549982:internal-dpl-injected-dummy-sink]: [17:16:07][ERROR] ConsumeExisting
[549960:mch-entropy-encoder]: [17:16:07][ERROR] sink:false action:consume
[549928:gpu-reconstruction_t1]: [17:16:07][ERROR] sink:false action:consume
[549969:itstpc-track-matcher_t0]: [17:16:07][ERROR] Consume:true
[549939:mft-stf-decoder]: [17:16:07][ERROR] sink:false action:consume
[549982:internal-dpl-injected-dummy-sink]: [17:16:27][ERROR] Invoked
[549982:internal-dpl-injected-dummy-sink]: [17:16:27][ERROR] sink:false action:consumeExisting
[549982:internal-dpl-injected-dummy-sink]: [17:16:27][ERROR] ConsumeExisting
[549960:mch-entropy-encoder]: [17:16:27][ERROR] sink:false action:consume
[549939:mft-stf-decoder]: [17:16:27][ERROR] sink:false action:consume
[549961:mft-entropy-encoder]: [17:16:28][ERROR] sink:false action:consume
[549950:its-tracker_t1]: [17:16:28][ERROR] sink:false action:consume
[549972:TRDTRACKLETTRANSFORMER_t0]: [17:16:28][ERROR] Consume:true
[549982:internal-dpl-injected-dummy-sink]: [17:16:28][ERROR] Invoked
[549982:internal-dpl-injected-dummy-sink]: [17:16:28][ERROR] sink:false action:consumeExisting
[549982:internal-dpl-injected-dummy-sink]: [17:16:28][ERROR] ConsumeExisting
[549972:TRDTRACKLETTRANSFORMER_t0]: [17:16:28][ERROR] sink:false action:consume
[549969:itstpc-track-matcher_t0]: [17:16:28][ERROR] sink:false action:consume
[549979:tof-matcher]: [17:16:28][ERROR] Consume:true
[549979:tof-matcher]: [17:16:28][ERROR] sink:false action:consume
[549982:internal-dpl-injected-dummy-sink]: [17:16:28][ERROR] Invoked
[549982:internal-dpl-injected-dummy-sink]: [17:16:28][ERROR] sink:false action:consumeExisting
[549982:internal-dpl-injected-dummy-sink]: [17:16:28][ERROR] ConsumeExisting
[549980:trd-globaltracking_ITS-TPC_]: [17:16:29][ERROR] sink:false action:consume
```
Is this just debug output that was left, or is something wrong. The chain seems to run normally with all detectors,

**Comment by davidrohr** [comment]
Actually, I was too early. The FST eventually died with backpressure, and the internal-dummy-sink segfaulted with this stack trace:
```
[549982:internal-dpl-injected-dummy-sink]: ===========================================================      
[549982:internal-dpl-injected-dummy-sink]: #6  boost::unordered::unordered_map<unsigned short, fair::mq::shmem::SegmentInfo, boost::hash<unsigned short>, std::equal_to<unsigned short>, boost::interprocess::allocator<std::pair<unsigned short const, fair::mq::shmem::SegmentInfo>, boost::interprocess::segment_manager<char, boost::interprocess::rbt
ree_best_fit<boost::interprocess::mutex_family, boost::interprocess::offset_ptr<void, long, unsigned long, 0ul>, 0ul>, boost::interprocess::iset_index> > >::at (k=
[549982:internal-dpl-injected-dummy-sink]: 0x7ffdf7c2b5fc: 1, this=0x7f3471400178) at /home/drohr/alice-4.3/sw/slc8_x86-64/boost/v1.75.0-local2/include/boost/unordered/unordered_map.hpp:1878                                                                                                                                                           
[549982:internal-dpl-injected-dummy-sink]: #7  fair::mq::shmem::Manager::GetSegment (this=0x1016930, id=1) at /home/drohr/alice-4.3/sw/SOURCES/FairMQ/v1.4.42/v1.4.42/fairmq/shmem/Manager.h:669                                                                                                                                                         
[549982:internal-dpl-injected-dummy-sink]: #8  0x00007f347e565bc5 in fair::mq::shmem::Message::Deallocate (this=0x12d1320) at /home/drohr/alice-4.3/sw/SOURCES/FairMQ/v1.4.42/v1.4.42/fairmq/shmem/Message.h:337                                                                                                                                         
[549982:internal-dpl-injected-dummy-sink]: #9  fair::mq::shmem::Message::CloseMessage (this=0x12d1320) at /home/drohr/alice-4.3/sw/SOURCES/FairMQ/v1.4.42/v1.4.42/fairmq/shmem/Message.h:378                                                                                                                                                             
[549982:internal-dpl-injected-dummy-sink]: #10 0x00007f347e565ee7 in fair::mq::shmem::Message::~Message (this=0x12d1320, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/SOURCES/FairMQ/v1.4.42/v1.4.42/fairmq/shmem/Message.h:310                                                                                                                
[549982:internal-dpl-injected-dummy-sink]: #11 fair::mq::shmem::Message::~Message (this=0x12d1320, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/SOURCES/FairMQ/v1.4.42/v1.4.42/fairmq/shmem/Message.h:310                                                                                                                                      
[549982:internal-dpl-injected-dummy-sink]: #12 0x00007f347fd957bf in std::default_delete<fair::mq::Message>::operator() (__ptr=<optimized out>, this=0x12f07f0) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/unique_ptr.h:79                                                                      
[549982:internal-dpl-injected-dummy-sink]: #13 std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> >::~unique_ptr (this=0x12f07f0, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/unique_ptr.h:361                                                 
[549982:internal-dpl-injected-dummy-sink]: #14 std::_Destroy<std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> > > (__pointer=0x12f07f0) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/stl_construct.h:140                                                                 
[549982:internal-dpl-injected-dummy-sink]: #15 std::_Destroy_aux<false>::__destroy<std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> >*> (__last=<optimized out>, __first=0x12f07f0) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/stl_construct.h:152                     
[549982:internal-dpl-injected-dummy-sink]: #16 std::_Destroy<std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> >*> (__last=<optimized out>, __first=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/stl_construct.h:185                                     
[549982:internal-dpl-injected-dummy-sink]: #17 std::_Destroy<std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> >*, std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> > > (__last=0x12f09b0, __first=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c
++/10.2.0/bits/alloc_traits.h:738                                                                                                                                                                                                                                                                                                                         
[549982:internal-dpl-injected-dummy-sink]: #18 std::vector<std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> >, std::allocator<std::unique_ptr<fair::mq::Message, std::default_delete<fair::mq::Message> > > >::~vector (this=0xd435d0, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-a
lice2-local1/include/c++/10.2.0/bits/stl_vector.h:680                                                                                                                                                                                                                                                                                                    
[549982:internal-dpl-injected-dummy-sink]: #19 fair::mq::Parts::~Parts (this=0xd435d0, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/FairMQ/v1.4.42-local4/include/fairmq/Parts.h:35                                                                                                                                                
[549982:internal-dpl-injected-dummy-sink]: #20 o2::framework::InputChannelInfo::~InputChannelInfo (this=0xd435c0, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/SOURCES/O2/dev/0/Framework/Core/include/Framework/ChannelInfo.h:36                                                                                                              
[549982:internal-dpl-injected-dummy-sink]: #21 std::_Destroy<o2::framework::InputChannelInfo> (__pointer=0xd435c0) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/stl_construct.h:140                                                                                                               
[549982:internal-dpl-injected-dummy-sink]: #22 std::_Destroy_aux<false>::__destroy<o2::framework::InputChannelInfo*> (__last=<optimized out>, __first=0xd435c0) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/stl_construct.h:152                                                                  
[549982:internal-dpl-injected-dummy-sink]: #23 std::_Destroy<o2::framework::InputChannelInfo*> (__last=<optimized out>, __first=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/stl_construct.h:185                                                                                 
[549982:internal-dpl-injected-dummy-sink]: #24 std::_Destroy<o2::framework::InputChannelInfo*, o2::framework::InputChannelInfo> (__last=0xd43800, __first=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/alloc_traits.h:738                                                        
[549982:internal-dpl-injected-dummy-sink]: #25 std::vector<o2::framework::InputChannelInfo, std::allocator<o2::framework::InputChannelInfo> >::~vector (this=0x1250970, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/stl_vector.h:680                                  
[549982:internal-dpl-injected-dummy-sink]: #26 o2::framework::DeviceState::~DeviceState (this=0x1250970, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/SOURCES/O2/dev/0/Framework/Core/include/Framework/DeviceState.h:43                                                                                                                       
[549982:internal-dpl-injected-dummy-sink]: #27 std::default_delete<o2::framework::DeviceState>::operator() (__ptr=0x1250970, this=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/unique_ptr.h:85                                                                                   
[549982:internal-dpl-injected-dummy-sink]: #28 0x00007f347fdb2606 in std::default_delete<o2::framework::DeviceState>::operator() (__ptr=<optimized out>, this=0x7ffdf7c2c4d8) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/unique_ptr.h:79                                                        
[549982:internal-dpl-injected-dummy-sink]: #29 std::unique_ptr<o2::framework::DeviceState, std::default_delete<o2::framework::DeviceState> >::~unique_ptr (this=0x7ffdf7c2c4d8, __in_chrg=<optimized out>) at /home/drohr/alice-4.3/sw/slc8_x86-64/GCC-Toolchain/v10.2.0-alice2-local1/include/c++/10.2.0/bits/unique_ptr.h:361                           
[549982:internal-dpl-injected-dummy-sink]: #30 doChild (argc=<optimized out>, argv=0x7ffdf7c2c860, serviceRegistry=..., runningWorkflow=..., ref=..., processingPolicies=..., defaultDriverClient=..., loop=<optimized out>) at /home/drohr/alice-4.3/sw/SOURCES/O2/dev/0/Framework/Core/src/runDataProcessing.cxx:1054                                  
[549982:internal-dpl-injected-dummy-sink]: #31 0x00007f347fdb36c4 in runStateMachine (workflow=..., workflowInfo=..., previousDataProcessorInfos=..., commandInfo=..., driverControl=..., driverInfo=..., metricsInfos=..., varmap=..., frameworkId=...) at /home/drohr/alice-4.3/sw/SOURCES/O2/dev/0/Framework/Core/src/runDataProcessing.cxx:1565       
[549982:internal-dpl-injected-dummy-sink]: #32 0x00007f347fdbd1a7 in doMain (argc=argc
[549982:internal-dpl-injected-dummy-sink]: entry=105, argv=argv                                                                                                                                                                                                                                                                                          
[549982:internal-dpl-injected-dummy-sink]: entry=0x7ffdf7c310f8, workflow=..., channelPolicies=..., completionPolicies=..., dispatchPolicies=..., resourcePolicies=..., currentWorkflowOptions=..., configContext=...) at /home/drohr/alice-4.3/sw/SOURCES/O2/dev/0/Framework/Core/src/runDataProcessing.cxx:2552                                        
[549982:internal-dpl-injected-dummy-sink]: #33 0x000000000040ea11 in mainNoCatch (argc=105, argv=0x7ffdf7c310f8) at /home/drohr/alice-4.3/sw/SOURCES/O2/dev/0/Framework/Core/include/Framework/runDataProcessing.h:201                                                                                                                                   
[549982:internal-dpl-injected-dummy-sink]: #34 0x000000000040b6e2 in main (argc=105, argv=0x7ffdf7c310f8) at /home/drohr/alice-4.3/sw/SOURCES/O2/dev/0/Framework/Core/include/Framework/runDataProcessing.h:219                                                                                                                                          
[549982:internal-dpl-injected-dummy-sink]: =========================================================== 
```

**Comment by ktf** [comment]
Do you have anything which produces sampled outputs in the chain? If yes, we need to mark those as `Lifetime::QA`.

**Comment by davidrohr** [comment]
no, shouldn't have that

**Comment by davidrohr** [comment]
There seems to be still this error message as debug: [1101455:trd-entropy-encoder]: [09:33:31][ERROR] Consume:true

**Comment by davidrohr** [comment]
OK, this is working in the FST now, at least it is not producing more problems that there are before this PR.
But could we add a command line option that enables the new behavior, such by default nothing changes. I think it is too invasive to deploy it fast otherwise.

**Comment by davidrohr** [comment]
This has a conflicting file after you merged the other one.
I'd propose we wait for the full CI to finish before you rebase, then I can test once on the EPN in the FST and then we just merge.

**Comment by ktf** [comment]
I have added an option (--timeframes-rate-limit) which should disable it by default.

**Comment by ktf** [comment]
This is the closest I get to working. The remaining issue seems to be I am not able to set the channel config properly for the `internal-dpl-injected-dummy-sink` and therefore the sink is not pushing the metric to the input proxy which then waits forever.

**Comment by ktf** [comment]
Ok, I somehow managed to hardcode values and this works, i.e. I see the metric being reported and the throttling logic does its job. We really need to rework how command line options are constructed, especially for channel config.

**Comment by ktf** [comment]
@pzhristov @TimoWilken now looks like both o2-cs8 and o2-dataflow-cs8 are failing due to picking up the wrong clang version. Weird because the alidist tests are still all green.

**Comment by ktf** [comment]
@davidrohr Actually, this one is the one we care about for the FST.

**Comment by davidrohr** [comment]
Could you fix clang-format and
```36/95 Test  #50: test_Framework_test_FrameworkDataFlowToDDS ............***Failed    0.07 sec```

Now testing this in full system test

**Review by davidrohr — DISMISSED** [review]
Works in the full system test. Let's get the ctest and formatting fixed and then get this merged!

**Comment by ktf** [comment]
@davidrohr this should compile. Not yet tested.

**Review by davidrohr — DISMISSED** [review]
From what I see, it doesn't disturb the processing with default settings, and the TF rate limiting works in my test
LGTM ASAP

**Comment by davidrohr** [comment]
errors unrelated, merging

**Comment by ktf** [comment]
This seems to introduce some sporadic race condition on end of stream. Looking into it.

**Pull request merged**
